src/openrct2/world/map.c

```diff
@@ -1846,40 +1886,72 @@ static money32 raise_land(sint32 flags, sint32 x, sint32 y, sint32 z, sint32 ax,
 static money32 lower_land(sint32 flags, sint32 x, sint32 y, sint32 z, sint32 ax, sint32 ay, sint32 bx, sint32 by, sint32 selectionType)
 {
     money32 cost = 0;
+    sint32  corners[2] = { MAP_SELECT_TYPE_EDGE_0, MAP_SELECT_TYPE_EDGE_0 };
 
-    if ((flags & GAME_COMMAND_FLAG_APPLY) && gGameCommandNestLevel == 1) {
-        audio_play_sound_at_location(SOUND_PLACE_ITEM, x, y, z);
+    if (selectionType >= MAP_SELECT_TYPE_EDGE_0 && selectionType <= MAP_SELECT_TYPE_EDGE_3) 
+    {
+        sint32 diff = EDGE_TO_CORNERS_ROTATION_OFFSET;
+        //we need to transform edge - check double corner operations; diff cannot be explicit as define due to some optimalisations (different result)
+        //extract corners from edge:
+        corners[0] = map_correct_corners(selectionType - diff);
+        corners[1] = map_correct_corners(corners[0] - 1);
+    }
+    else
+    {
+        corners[0] = selectionType;
     }
 
-    if (selectionType < 0 || selectionType >= 5)
+    if (corners[0] < MAP_SELECT_TYPE_CORNER_0 || corners[0] > MAP_SELECT_TYPE_FULL ||
+    corners[0] >= countof(map_element_raise_styles))
     {
-        log_warning("Improper selection type %d", selectionType);
+        log_warning("Invalid selection type %d for raising land", selectionType);
         return MONEY32_UNDEFINED;
     }
 
+    if ((flags & GAME_COMMAND_FLAG_APPLY) && gGameCommandNestLevel == 1) 
+    {
+        audio_play_sound_at_location(SOUND_PLACE_ITEM, x, y, z);
+    }
+
     uint8 max_height = map_get_highest_land_height(ax, bx, ay, by);
 
-    for (sint32 yi = ay; yi <= by; yi += 32) {
-        for (sint32 xi = ax; xi <= bx; xi += 32) {
-            rct_map_element *map_element = map_get_surface_element_at(xi / 32, yi / 32);
-            if (map_element != NULL) {
-                uint8 height = map_element->base_height;
-                if (map_element->properties.surface.slope & 0xF)
-                    height += 2;
-                if (map_element->properties.surface.slope & 0x10)
-                    height += 2;
-                if (height >= max_height) {
-                    height =  map_element->base_height;
-                    uint8 newStyle = map_element_lower_styles[selectionType][map_element->properties.surface.slope & MAP_ELEMENT_SLOPE_MASK];
-                    if (newStyle & 0x20) { // needs to be lowered
-                        height -= 2;
-                        newStyle &= ~0x20;
+    for (uint8 cornerIndex = 0; cornerIndex < 2 &&
+    corners[cornerIndex] >= MAP_SELECT_TYPE_CORNER_0 &&
+    corners[cornerIndex] <= MAP_SELECT_TYPE_FULL;
+    cornerIndex++)
+    {
+        for (sint32 yi = ay; yi <= by; yi += 32) 
```

Comment:
I find `yi`, `ay`, `by` as variable names quite vague. Could you rename them to something more descriptive?